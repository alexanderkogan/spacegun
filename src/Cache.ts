export class Cache<S, T> {
    private data: Map<S, { value: T, age: number }> = new Map()

    public constructor(
        public maxAgeSeconds: number,
        public maxSize: number = 100
    ) { }

    public get(key: S): T | undefined {
        const entry = this.data.get(key)
        if (entry !== undefined && Date.now() - entry.age < this.maxAgeSeconds * 1000) {
            return entry.value
        }
        return undefined
    }

    public set(key: S, value: T): void {
        this.data.set(key, { value, age: Date.now() })
    }

    public async calculate(key: S, valueProvider: () => Promise<T>): Promise<T> {
        let value = this.get(key)
        if (value === undefined) {
            value = await valueProvider()
            this.set(key, value)
        }
        return value
    }
}
