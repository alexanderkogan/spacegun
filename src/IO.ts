import * as readline from 'readline'

export class IO {
    public constructor(
        public readonly stdin: NodeJS.ReadableStream = process.stdin,
        public readonly stdout: NodeJS.WritableStream = process.stdout
    ) {}

    public choose<T>(question: string, options: T[]): Promise<T> {
        return new Promise((resolve, reject) => {
            this.readlineContext(question, answer => {
                let index
                try {
                    index = Number.parseInt(answer)
                } catch (error) {}
                if (index === undefined || index < 0 || index >= options.length) {
                    reject(new Error(`${answer} is not in the valid range`))
                } else {
                    resolve(options[index])
                }
            })
        })
    }

    public chooseMultiple<S, T>(question: string, options1: S[], options2: T[]): Promise<{ first: boolean; result: S | T }> {
        return new Promise((resolve, reject) => {
            this.readlineContext(question, answer => {
                let index
                try {
                    index = Number.parseInt(answer)
                } catch (error) {}
                if (index === undefined || index < 0 || index >= options1.length + options2.length) {
                    reject(new Error(`${answer} is not in the valid range`))
                } else if (index >= options1.length) {
                    resolve({ first: false, result: options2[index - options1.length] })
                } else {
                    resolve({ first: true, result: options1[index] })
                }
            })
        })
    }

    public expect(question: string, expected: string): Promise<boolean> {
        return new Promise(resolve => {
            this.readlineContext(question, answer => resolve(answer === expected))
        })
    }

    public error(e: Error | string) {
        console.error(e)
    }

    public out(text: string) {
        console.log(text)
    }

    private readlineContext(question: string, callback: (answer: string) => void) {
        const r = readline.createInterface({
            input: this.stdin,
            output: this.stdout,
        })
        r.question(question, answer => {
            callback(answer)
            r.close()
        })
    }
}
