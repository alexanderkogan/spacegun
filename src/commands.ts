import { ClusterRepository } from "@/cluster/ClusterRepository"
import { Deployment } from "@/cluster/model/Deployment"

import { ImageRepository } from "@/images/ImageRepository"

import { pad } from "@/pad"
import chalk from "chalk"

import * as ora from "ora"
import { Question } from "@/Question"

export type Command = "pods" | "images" | "deployments" | "deploy" | "scalers" | "help"

async function load<T>(p: Promise<T>): Promise<T> {
    const progress = ora()
    progress.start("loading")
    const result = await p
    progress.stop()
    return result
}

async function foreachCluster(clusterRepository: ClusterRepository, f: (clusterRepository: ClusterRepository, cluster: string) => void) {
    for (const cluster of clusterRepository.clusters) {
        console.log("")
        console.log(chalk.underline.bold(pad(`${cluster}`)))
        await f(clusterRepository, cluster)
    }
}

async function podsCommand(clusterRepository: ClusterRepository, cluster: string) {
    const pods = await load(clusterRepository.pods(cluster))
    console.log(chalk.bold(pad("pod name", 5) + pad("tag name", 5) + pad("starts", 1), pad("status", 1)))
    pods.forEach(pod => {
        let restartText
        if (pod.restarts === undefined) {
            restartText = chalk.bold.cyan(pad("na!", 1))
        } else if (pod.restarts > 30) {
            restartText = chalk.bold.cyan(pad(pod.restarts.toString() + "!", 1))
        } else if (pod.restarts > 10) {
            restartText = chalk.bold.magenta(pad(pod.restarts.toString(), 1))
        } else {
            restartText = pad(pod.restarts.toString(), 1)
        }
        const statusText = pod.ready ? chalk.bold.magenta(pad("up", 1)) : chalk.bold.cyan(pad("down!", 1))
        let tagText
        if (pod.image === undefined) {
            tagText = chalk.bold.magenta(pad("missing", 5))
        } else {
            tagText = pad(pod.image.tag, 5)
        }
        console.log(pad(pod.name, 5) + tagText + restartText + statusText)
    })
}

async function imagesCommand(imageRepository: ImageRepository) {
    const images = await imageRepository.images()
    images.forEach(image =>
        console.log(image)
    )
}

function logDeploymentHeader() {
    console.log(chalk.bold(pad("deployment name", 5) + pad("tag name", 7)))
}

function logDeployment(deployment: Deployment) {
    let tagText
    if (deployment.image === undefined) {
        tagText = chalk.bold.magenta(pad("missing", 7))
    } else {
        tagText = pad(deployment.image.tag, 7)
    }
    console.log(pad(deployment.name, 5) + tagText)
}

async function deployementsCommand(clusterRepository: ClusterRepository, cluster: string) {
    const deployments = await load(clusterRepository.deployments(cluster))
    logDeploymentHeader()
    deployments.forEach(deployment => {
        logDeployment(deployment)
    })
}

async function deployCommand(clusterRepository: ClusterRepository, imageRepository: ImageRepository) {
    const question = new Question()
    try {
        console.log("Choose the target cluster")
        clusterRepository.clusters.forEach((cluster, index) => {
            console.log(chalk.bold.cyan(index.toString()) + ": " + pad(cluster, 5))
        })
        const targetCluster = await question.choose('> ', clusterRepository.clusters)

        const deployments = await load(clusterRepository.deployments(targetCluster))

        console.log("Choose the target deployment")
        deployments.forEach((deployment, index) => {
            console.log(chalk.bold.cyan(index.toString()) + ": " + pad(deployment.name, 5))
        })
        const targetDeployment = await question.choose('> ', deployments)

        const versions = await imageRepository.versions(targetDeployment.image!.name)
        versions.sort((a, b) => b.lastUpdated.getTime() - a.lastUpdated.getTime())
        console.log("Choose the target image")
        versions.forEach((image, index) => {
            if (image.tag === targetDeployment.image!.tag) {
                console.log(chalk.italic.magenta(index.toString()) + ": " + chalk.italic.magenta(pad(image.tag, 5)))
            } else {
                console.log(chalk.bold.cyan(index.toString()) + ": " + pad(image.tag, 5))
            }
        })
        const targetImage = await question.choose('> ', versions)

        console.log("deploy " + chalk.cyan(targetImage.url) + " into " + chalk.cyan(targetCluster + "::" + targetDeployment.name))
        console.log("Answer `yes` to apply.")
        const userAgrees = await question.expect('> ', "yes")
        if (userAgrees) {
            const updated = await clusterRepository.updateDeployment(targetCluster, targetDeployment, targetImage)
            logDeploymentHeader()
            logDeployment(updated)
        }
    } catch (error) {
        console.error(error)
    } finally {
        question.close()
    }
}

async function scalersCommand(clusterRepository: ClusterRepository, cluster: string) {
    const scalers = await load(clusterRepository.scalers(cluster))
    console.log(chalk.bold(pad("scaler name", 5) + pad("replication", 7)))
    console.log(chalk.bold(pad("", 5) + pad("current", 3) + pad("minimum", 2) + pad("maximum", 2)))
    scalers.forEach(scaler => {
        let line = pad(scaler.name, 5)
        let currentText = scaler.replicas.current.toString()
        if (scaler.replicas.current < scaler.replicas.minimum) {
            currentText = chalk.bold.cyan(pad(currentText + "!", 3))
        } else if (scaler.replicas.current >= scaler.replicas.maximum) {
            currentText = chalk.bold.magenta(pad(currentText, 3))
        } else {
            currentText = pad(currentText, 3)
        }
        line += currentText
        line += pad(scaler.replicas.minimum.toString(), 2) + pad(scaler.replicas.maximum.toString(), 2)
        console.log(line)
    })
}

export function printHelp(clusterRepository?: ClusterRepository, imageRepository?: ImageRepository, invalidConfig: boolean = false) {
    const b = chalk.blue;
    const m = chalk.magenta;
    const c = chalk.cyan;
    const CLI_TITLE = chalk.bold.underline('Spacegun-CLI')
    const CLI_DESCRIPTION = 'Space age deployment manager'
    const CLI_USAGE = 'Usage: \`spacegun <command> [options ...]\`'

    const HELP_HEADER = `
        ${b('/\\')} ${c('*')}    
       ${b('/__\\')}     ${CLI_TITLE}   ${b('version 0.0.8')}
      ${b('/\\  /')}
     ${b('/__\\/')}      ${CLI_DESCRIPTION}
    ${b('/\\')}  ${m('/\\')}     
   ${b('/__\\')}${m('/__\\')}     ${CLI_USAGE}
  ${b('/\\')}  ${m('/')}    ${m('\\')}
`
    console.log(HELP_HEADER)

    if (invalidConfig) {
        console.log(c('no configuration file found!'))
        console.log(c('A config.yml containing the following line might be sufficient'))
        console.log(b('docker: https://your.docker.registry/'))
        console.log("")
    }
    if (clusterRepository !== undefined) {
        console.log('configured clusters: ' + m(clusterRepository.clusters.join(", ")))
    } else {
        console.log(c('no clusters configured! (such as your kubernetes)'))
    }
    if (imageRepository !== undefined) {
        console.log('configured image endpoint: ' + m(imageRepository.endpoint))
    } else {
        console.log(c('no image registry configured! (such as your docker registry)'))
    }

    console.log('')
    console.log(chalk.bold.underline('Available Commands'))
    console.log(pad("pods", 2) + chalk.bold(pad("a summary of all pods of all known clusters", 10)))
    console.log(pad("images", 2) + chalk.bold(pad("a list of all images in the docker registry", 10)))
    console.log(pad("deployments", 2) + chalk.bold(pad("a summary of all deployements of all known clusters", 10)))
    console.log(pad("deploy", 2) + chalk.bold(pad("opens an interactive dialog to deploy an image", 10)))
    console.log(pad("scalers", 2) + chalk.bold(pad("a summary of all scalers of all known clusters", 10)))
    console.log(pad("help", 2) + chalk.bold(pad("renders this summary", 10)))
    console.log('')
    console.log(chalk.bold.underline('Available Options'))
    console.log(pad("config, c", 2) + chalk.bold(pad("path to the config.yml. Default: `config.yml`", 10)))
}

const commands: { [k in Command]: (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => Promise<void> } = {
    "pods": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, podsCommand)
    },
    "images": async ({ }: ClusterRepository, imageRepository: ImageRepository) => {
        imagesCommand(imageRepository)
    },
    "deployments": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, deployementsCommand)
    },
    "scalers": async (clusterRepository: ClusterRepository, { }: ImageRepository) => {
        foreachCluster(clusterRepository, scalersCommand)
    },
    "deploy": async (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => {
        deployCommand(clusterRepository, imageRepository)
    },
    "help": async (clusterRepository: ClusterRepository, imageRepository: ImageRepository) => {
        printHelp(clusterRepository, imageRepository)
    }
}

export { commands }
