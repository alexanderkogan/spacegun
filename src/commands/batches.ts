import chalk from 'chalk'

import * as clusterModule from '../cluster/ClusterModule'

import { CommandFn } from '.'
import { load, foreachNamespace, foreachCluster } from './helpers'

import { call } from '../dispatcher'
import { pad } from '../pad'
import { IO } from '../IO'
import { Options } from '../options'

import { Batch } from '../cluster/model/Batch'

export const batchesCommand: CommandFn = async (options: Options, io: IO) =>
    foreachCluster(options, io, (options, io, cluster) => foreachNamespace(options, io, cluster, batches))

async function batches(io: IO, cluster: string, namespace?: string) {
    const batches = await load(call(clusterModule.batches)({ cluster, namespace }))
    logBatchHeader(io)
    batches.forEach(batch => {
        logBatch(io, batch)
    })
}

export function logBatchHeader(io: IO) {
    io.out(chalk.bold(pad('deployment name', 5) + pad('image url', 7)))
}

export function logBatch(io: IO, deployment: Batch) {
    let urlText
    if (deployment.image === undefined) {
        urlText = chalk.bold.magenta(pad('missing', 7))
    } else {
        urlText = pad(deployment.image.url, 7)
    }
    io.out(pad(deployment.name, 5) + urlText)
}
