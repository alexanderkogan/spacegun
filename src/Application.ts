import { load as loadConfig, Config } from "@/config"
import { commands, printHelp } from "@/commands"
import { parse } from "@/options"

import { Layers } from "@/dispatcher/model/Layers"
import { run as runDispatcher } from "@/dispatcher"

import { IO } from "@/IO"
import { CronRegistry } from "@/crons/CronRegistry"

import { GitConfigRepository, fromConfig as gitRepoFromConfig } from "@/config/git/GitConfigRepository"
import { fromConfig as configRepoFromConfig } from "@/config/filesystem/FilesystemConfigRepository";

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"
import { SlackEventRepository } from "@/events/slack/SlackEventRepository"

import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initEvents } from "@/events/EventModule"
import { init as initImages } from "@/images/ImageModule"
import { init as initConfig } from "@/config/ConfigModule"
import { init as initJobs } from "@/jobs/JobsModule"
import { init as initViews } from "@/views"

import { Options } from "@/options"

export class Application {

    public static create(): Application {
        const io = new IO()
        const crons = new CronRegistry()
        const options = parse()
        return new Application(io, crons, options)
    }

    public constructor(
        public readonly io: IO,
        public readonly crons: CronRegistry,
        public readonly options: Options
    ) { }

    public async run() {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)

            runDispatcher(config.server.host, config.server.port)
            if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
                await commands[this.options.command](this.io)
            }
        } catch (e) {
            printHelp(this.io, e)
        }
    }

    public async checkForConfigChange(git: GitConfigRepository) {
        const hasNewConfig = await git.hasNewConfig()
        if (hasNewConfig) {
            await git.fetchNewConfig()
            this.crons.removeAllCrons()
            this.reload()
        }
    }

    private reload(): void {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)
        } catch (e) {
            this.io.out(`could not reload config ${e.message}`)
        }
    }

    private initialize(config: Config) {
        const gitRepo = gitRepoFromConfig(config)
        if (gitRepo !== undefined) {
            this.crons.register(
                "config-reload",
                config.git!.cron,
                () => this.checkForConfigChange(gitRepo)
            )
            initConfig(gitRepo)
        } else {
            const fileSystemRepo = configRepoFromConfig(config)
            initConfig(fileSystemRepo)
        }

        initViews(config)
        initEvents([
            SlackEventRepository.fromConfig(config.slack)
        ])
        initCluster(KubernetesClusterRepository.fromConfig(config.kube, config.namespaces))
        initImages(DockerImageRepository.fromConfig(config.docker))
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs, this.crons))
    }
}
