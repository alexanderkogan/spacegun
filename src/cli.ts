import { parse } from "./options"
import { KubernetesClusterProvider } from './cluster/kubernetes/KubernetesCluster'
import { commands } from "./commands"

(async () => {
    const options = parse()
    const cluster = new KubernetesClusterProvider(options.kube)
    try {
        await commands[options.command](cluster)
    } catch (error) {
        console.error(`Command ${options.command} failed: ${error}`)
    }
})()
