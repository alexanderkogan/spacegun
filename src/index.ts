import { Layers } from "@/dispatcher/model/Layers"

import { load as loadConfig } from "@/config"
import { parse } from "@/options"
import { run as runDispatcher } from "@/dispatcher"
import { commands, printHelp } from "@/commands"
import { IO } from "@/IO"

import { init as initJobs } from "@/jobs/JobsModule"
import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initImages } from "@/images/ImageModule"

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"

const options = parse()
const config = loadConfig(options.config)
const io = new IO()
try {
    if (config instanceof Error) {
        printHelp(io, true)
    } else {
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs))
        initCluster(KubernetesClusterRepository.fromConfig(config.kube))
        initImages(DockerImageRepository.fromConfig(config.docker))

        runDispatcher(config.server.host, config.server.port)
        if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
            commands[options.command](io)
        }
    }
} catch (e) {
    console.error(e)
}
