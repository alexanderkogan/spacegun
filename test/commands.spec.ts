import { IO } from "../src/IO"

const mockDispatched = jest.fn()
const mockDispatchFn = jest.fn()
jest.mock("../src/dispatcher/index", () => ({
    get: (moduleName: string, procedureName: string) => {
        mockDispatched(moduleName, procedureName)
        return mockDispatchFn
    },
    call: (request: any) => {
        mockDispatched(request.module, request.procedure)
        return mockDispatchFn
    },
    add: () => { },
    path: () => ""
}))

import { commands } from "../src/commands"
import * as moment from "moment"

describe("commands", () => {
    beforeEach(() => {
        jest.resetAllMocks()
    })

    describe(commands.images.name, () => {

        it("calls the images backend and prints images", async () => {

            // given
            const image = { imageName: "someImage" }
            mockDispatchFn.mockReturnValue([image])

            // when
            await commands.images(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(1)
            expect(mockDispatched).toBeCalledWith("images", "list")
        })
    })

    describe(commands.namespaces.name, () => {

        it("calls the namespaces backend and prints namespace", async () => {
            // given
            mockDispatchFn.mockReturnValue(["namespace1", "namespace2"])

            // when
            await commands.namespaces(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(3)
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
        })
    })

    describe(commands.pipelines.name, () => {

        it("calls the jobs backend and prints pipelines", async () => {
            // given
            const pipeline = { name: "1->2", cluster: "cluster2", steps: [], start: "" }
            mockDispatchFn.mockReturnValue([pipeline])

            // when
            await commands.pipelines(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(1)
            expect(mockDispatched).toBeCalledWith("jobs", "pipelines")
        })
    })

    describe(commands.pods.name, () => {

        it("calls the pods backend for each cluster", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(["cluster1", "cluster2"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce([{ name: "service1", age: "2 days" }])
                .mockReturnValueOnce(["service1"])
                .mockReturnValueOnce([{ name: "service1", age: "2 days" }])

            // when
            await commands.pods(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(5)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "pods")
        })

        it("prints out the pods correctly", async () => {
            // given
            const withAgeOf = function (amount: moment.DurationInputArg1, unit: moment.DurationInputArg2): number {
                let now = moment()
                now.subtract(amount, unit)
                return now.valueOf()
            }
            mockDispatchFn
                .mockReturnValueOnce(["cluster1"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce([
                    {
                        name: "service1",
                        image: undefined,
                        restarts: undefined,
                        ready: true,
                        creationTimeMS: withAgeOf(1, "day")
                    },
                    {
                        name: "service2",
                        image: { url: "url1" },
                        restarts: 1,
                        ready: false,
                        creationTimeMS: withAgeOf(2, "days")
                    },
                    {
                        name: "service3",
                        image: undefined,
                        restarts: 11,
                        ready: true,
                        creationTimeMS: withAgeOf(3, "days")
                    },
                    {
                        name: "service4",
                        image: { url: "url2" },
                        restarts: 111,
                        ready: false,
                        creationTimeMS: withAgeOf(4, "days")
                    }
                ])
            let output: string[] = []
            const out = jest.fn((text: string) => {
                output.push(text)
                return text
            })
            const io = createIO({ out })

            // when
            await commands.pods(io)

            // then
            expect(output.concat("\n")).toMatchSnapshot()
        })
    })

    describe(commands.scalers.name, () => {

        it("calls the scalers backend for each cluster", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(["cluster1", "cluster2"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce([{ name: "scaler1", replicas: { current: 0, minimum: 1, maximum: 2 } }])
                .mockReturnValueOnce(["service1"])
                .mockReturnValueOnce([{ name: "scaler1", replicas: { current: 1, minimum: 1, maximum: 1 } }])

            // when
            await commands.scalers(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(5)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "scalers")
        })
    })

    describe(commands.deployments.name, () => {

        it("calls the pods deployments for each cluster", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(["cluster1", "cluster2"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce([{ name: "deployment1" }])
                .mockReturnValueOnce(["service1"])
                .mockReturnValueOnce([{ name: "deployment2" }])

            // when
            await commands.deployments(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(5)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "deployments")
        })
    })

    describe(commands.run.name, () => {

        it("runs a pipeline if user agrees", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce([{ name: "pipeline1" }, { name: "pipeline2" }])
                .mockReturnValueOnce({
                    name: "plan", deployments: [
                        { name: "deployment1", deployment: {}, image: {} }
                    ]
                })

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => true)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.run(io)

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(3)
            expect(mockDispatched).toBeCalledWith("jobs", "pipelines")
            expect(mockDispatched).toBeCalledWith("jobs", "plan")
            expect(mockDispatched).toBeCalledWith("jobs", "run")
        })

        it("does not run a pipeline if user disagrees", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce([{ name: "pipeline1" }, { name: "pipeline2" }])
                .mockReturnValueOnce({
                    name: "plan", deployments: [
                        { name: "deployment1", deployment: {}, image: {} }
                    ]
                })

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => false)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.run(io)

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(2)
            expect(mockDispatched).not.toBeCalledWith("jobs", "run")
        })
    })

    describe(commands.deploy.name, () => {
        const clusters = ["cluster1", "cluster2"]
        const namespaces: string[] = []
        const deployments = [{ name: "deployment1", image: { name: "image" } }]
        const tags = ["latest", "notLatest"]
        const image = { name: "image", tag: "latest", url: "image:latest:url" }

        it("deploys an image if user agrees", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(clusters)
                .mockReturnValueOnce(namespaces)
                .mockReturnValueOnce(deployments)
                .mockReturnValueOnce(tags)
                .mockReturnValueOnce(image)
                .mockReturnValueOnce({ name: "deployment1" })

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => true)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.deploy(io)

            // then
            expect(choose).toHaveBeenCalledTimes(3)
            expect(choose).toHaveBeenCalledWith("> ", clusters)
            expect(choose).toHaveBeenCalledWith("> ", deployments)
            expect(choose).toHaveBeenCalledWith("> ", tags)

            expect(mockDispatched).toHaveBeenCalledTimes(6)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "deployments")
            expect(mockDispatched).toBeCalledWith("images", "tags")
            expect(mockDispatched).toBeCalledWith("images", "image")
            expect(mockDispatched).toBeCalledWith("cluster", "updateDeployment")
        })

        it("asks user for namespace if there are any", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(clusters)
                .mockReturnValueOnce(["service1"])
                .mockReturnValueOnce(deployments)
                .mockReturnValueOnce(tags)
                .mockReturnValueOnce(image)
                .mockReturnValueOnce({ name: "deployment1" })

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => true)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.deploy(io)

            // then
            expect(choose).toHaveBeenCalledTimes(4)
            expect(choose).toHaveBeenCalledWith("> ", clusters)
            expect(choose).toHaveBeenCalledWith("> ", deployments)
            expect(choose).toHaveBeenCalledWith("> ", tags)
        })

        it("does not deploy an image if user disagrees", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(clusters)
                .mockReturnValueOnce(namespaces)
                .mockReturnValueOnce(deployments)
                .mockReturnValueOnce(tags)
                .mockReturnValueOnce(image)
                .mockReturnValueOnce({ name: "deployment1" })

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => false)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.deploy(io)

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(5)
            expect(mockDispatched).not.toBeCalledWith("cluster", "updateDeployment")
        })
    })

    describe(commands.pipelineSchedules.name, () => {

        it("prints schedules of a pipeline", async () => {
            // given
            const pipeline = { name: "1->2", cluster: "cluster2", steps: [], start: "" }
            const schedules = { lastRun: undefined, nextRuns: [] }
            mockDispatchFn
                .mockReturnValueOnce([pipeline])
                .mockReturnValueOnce(schedules)

            const choose = jest.fn().mockImplementation(({ }, b) => b[0])
            const expectFn = jest.fn().mockImplementation(() => true)
            const io = createIO({ choose, expect: expectFn })

            // when
            await commands.pipelineSchedules(io)

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(2)
            expect(mockDispatched).toBeCalledWith("jobs", "pipelines")
            expect(mockDispatched).toBeCalledWith("jobs", "schedules")
        })
    })

    describe(commands.help.name, () => {

        it("fetches cluster and image endpoint names", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(["cluster"])
                .mockReturnValueOnce("someEndpoint")

            // when
            await commands.help(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(2)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("images", "endpoint")
        })
    })

    describe(commands.snapshot.name, () => {

        it("creates snapshots", async () => {
            // given
            mockDispatchFn
                .mockReturnValueOnce(["cluster"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce({
                    deployments: [
                        { data: {}, name: "deployment1" },
                        { data: {}, name: "deployment2" }
                    ]
                })

            // when
            await commands.snapshot(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(5)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "takeSnapshot")
            expect(mockDispatched).toBeCalledWith("artifacts", "saveArtifact")
            expect(mockDispatched).toBeCalledWith("artifacts", "saveArtifact")
        })
    })

    describe(commands.apply.name, () => {

        it("applies snapshots", async () => {
            // given
            const deployments = [
                { name: "deployment1", image: { name: "image" } },
                { name: "deployment2", image: { name: "image" } }
            ]
            mockDispatchFn
                .mockReturnValueOnce(["cluster"])
                .mockReturnValueOnce([])
                .mockReturnValueOnce(deployments)
                .mockReturnValueOnce({})
                .mockReturnValueOnce({})

            // when
            await commands.apply(createIO())

            // then
            expect(mockDispatched).toHaveBeenCalledTimes(6)
            expect(mockDispatched).toBeCalledWith("cluster", "clusters")
            expect(mockDispatched).toBeCalledWith("cluster", "namespaces")
            expect(mockDispatched).toBeCalledWith("cluster", "deployments")
            expect(mockDispatched).toBeCalledWith("artifacts", "loadArtifact")
            expect(mockDispatched).toBeCalledWith("artifacts", "loadArtifact")
            expect(mockDispatched).toBeCalledWith("cluster", "applySnapshot")
        })
    })
})

function createIO(mocks: Partial<{ choose: jest.Mock<{}>, expect: jest.Mock<{}>, out: jest.Mock<{}> }> = {}): IO {
    const io = new IO()
    io.out = mocks.out || jest.fn()
    io.expect = mocks.expect || jest.fn()
    io.choose = mocks.choose || jest.fn()
    return io
}
