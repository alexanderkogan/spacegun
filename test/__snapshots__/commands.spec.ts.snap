// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands pods prints out the pods correctly 1`] = `
Array [
  "",
  "[1mpod name                                image url                               starts  status  age     [22m",
  "service1                                [1m[35mmissing                                 [39m[22m[1m[36mna!     [39m[22m[1m[35mup      [39m[22ma day   ",
  "service2                                url1                                    1       [1m[36mdown!   [39m[22m2 days  ",
  "service3                                [1m[35mmissing                                 [39m[22m[1m[35m11      [39m[22m[1m[35mup      [39m[22m3 days  ",
  "service4                                url2                                    [1m[36m111!    [39m[22m[1m[36mdown!   [39m[22m4 days  ",
  "
",
]
`;
